src/sim/init_signals.cc

```diff
@@ -213,29 +257,117 @@ externalProcessHandler(int sigtype)
     std::memcpy(full_payload, shm_ptr, shared_mem_size);
     full_payload[shared_mem_size - 1] = '\0';  // Ensure null-termination
 
-    std::cout << "Received signal from external process with payload: '"
-        << full_payload << "'" << std::endl;
+    DPRINTF(ExternalSignal, "Received signal from external "
+            "process with payload: '%s'\n", full_payload);
 
     // process payload json with string processing
     std::string full_payload_str = full_payload;
     std::map<std::string, std::string> payload_map;
 
-    //get the hypercall id
-    std::size_t payload_pos = 0;
-    std::string hypercall_id_str = extractStringFromJSON(full_payload_str,
-        "\"id\":", ",", payload_pos);
-    uint64_t hypercall_id = std::stoi(hypercall_id_str);
+    // Find ID field
+    std::size_t id_pos = full_payload_str.find("\"id\":");
+    if (id_pos == std::string::npos) {
+        munmap(shm_ptr, shared_mem_size);
+        close(shm_fd);
+        DPRINTF(ExternalSignal, "Error: No message ID found in payload\n");
+        return;
+    }
+
+    // Skip past "id":
+    id_pos += 5;
+
+    // Skip whitespace
+    while (id_pos < full_payload_str.length() && isspace(id_pos)) {
+        id_pos++;
+    }
+
+    // Find end of number (comma or closing brace)
+    std::size_t id_end = full_payload_str.find_first_of(",}", id_pos);
+    if (id_end == std::string::npos) {
+        munmap(shm_ptr, shared_mem_size);
+        close(shm_fd);
+        DPRINTF(ExternalSignal, "Error: Invalid ID format\n");
+        return;
+    }
+
+    std::string message_id_str = full_payload_str.substr(id_pos,
+                                                         id_end - id_pos);
+    // Trim whitespace
+    while (!message_id_str.empty() && isspace(message_id_str.back())) {
+        message_id_str.pop_back();
+    }
+
+    long long id_value = std::stoll(message_id_str);
+    if (id_value < 0) {
+        munmap(shm_ptr, shared_mem_size);
+        close(shm_fd);
+        panic("Error: Invalid ID format "
+              "- must be a valid non-negative 64-bit integer\n");
```

Comment:
I think a panic here is too much. This would kill the gem5 process which would be a harsh in the case someone accidentally types an invalid id".

I wouldn't want someone 10 days into a spec simulation accidentally killing their gem5 instance.